cmd/qat_plugin/kerneldrv/kerneldrv.go

```diff
@@ -0,0 +1,325 @@
+// Copyright 2018 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kerneldrv
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/go-ini/ini"
+	"github.com/pkg/errors"
+
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1beta1"
+	utilsexec "k8s.io/utils/exec"
+
+	"github.com/intel/intel-device-plugins-for-kubernetes/pkg/debug"
+	dpapi "github.com/intel/intel-device-plugins-for-kubernetes/pkg/deviceplugin"
+)
+
+const (
+	namespace = "qat.intel.com"
+)
+
+var (
+	uioRegex = regexp.MustCompile(`^uio[0-9]+$`)
+)
+
+type endpoint struct {
+	id        string
+	processes int
+}
+
+type section struct {
+	endpoints          []endpoint
+	cryptoEngines      int
+	compressionEngines int
+	pinned             bool
+}
+
+type driverConfig map[string]section
+
+func newDeviceSpec(devPath string) pluginapi.DeviceSpec {
+	return pluginapi.DeviceSpec{
+		HostPath:      devPath,
+		ContainerPath: devPath,
+		Permissions:   "rw",
+	}
+}
+
+func getDevTree(devfs string, config map[string]section) (dpapi.DeviceTree, error) {
+	devTree := dpapi.NewDeviceTree()
+
+	devFiles, err := ioutil.ReadDir(devfs)
+	if err != nil {
+		return devTree, errors.Wrapf(err, "Can't read %s", devfs)
+	}
+
+	devs := []pluginapi.DeviceSpec{
+		newDeviceSpec(path.Join(devfs, "qat_adf_ctl")),
+		newDeviceSpec(path.Join(devfs, "qat_dev_processes")),
+		newDeviceSpec(path.Join(devfs, "usdm_drv")),
+	}
+	for _, devFile := range devFiles {
+		fname := devFile.Name()
+
+		if uioRegex.MatchString(fname) {
```

Comment:
check the uio device belongs to QAT?

Reply:
I've updated the code to get the list of uio devices directly from sysfs, e.g. from `/sys/bus/pci/drivers/c6xx/0000:3d:00.0/uio/`